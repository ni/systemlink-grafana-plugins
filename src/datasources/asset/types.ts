import { FieldDTO } from '@grafana/data'
import { DataQuery } from '@grafana/schema'

export interface AssetMetadataQuery extends DataQuery {
  queryKind: AssetQueryType,
  workspace: string,
  minionIds: string[]
}

export interface AssetCalibrationForecastQuery extends DataQuery {
  queryKind: AssetQueryType,
  groupBy: string[]
}

export type AssetQuery = AssetCalibrationForecastQuery | AssetMetadataQuery;

export enum AssetQueryType {
  Metadata = "Metadata",
  CalibrationForecast = "Calibration Forecast",
}

export enum AssetQueryLabel {
  Metadata = "Metadata",
  CalibrationForecast = "Calibration Forecast",
}

export enum AssetCalibrationForecastGroupByType {
  Day = "DAY",
  Week = "WEEK",
  Month = "MONTH",
}

export enum AssetCalibrationForecastKey {
  Time = "Time",
}

export enum EntityType {
  Asset = "Asset",
  System = "System"
}

export interface AssetsResponse {
  assets: AssetModel[],
  totalCount: number
}

export interface AssetModel {
  modelName: string,
  modelNumber: number,
  serialNumber: string,
  vendorName: string,
  vendorNumber: number,
  busType: 'BUILT_IN_SYSTEM' | 'PCI_PXI' | 'USB' | 'GPIB' | 'VXI' | 'SERIAL' | 'TCP_IP' | 'CRIO' | 'SCXI' | 'CDAQ' | 'SWITCH_BLOCK' | 'SCC' | 'FIRE_WIRE' | 'ACCESSORY' | 'CAN' | 'SWITCH_BLOCK_DEVICE' | 'SLSC' | string,
  name: string,
  assetType: 'GENERIC' | 'DEVICE_UNDER_TEST' | 'FIXTURE' | 'SYSTEM' | string,
  firmwareVersion: string,
  hardwareVersion: string,
  visaResourceName: string,
  temperatureSensors: any[],
  supportsSelfCalibration: boolean,
  supportsExternalCalibration: boolean,
  customCalibrationInterval?: number,
  selfCalibration?: any,
  isNIAsset: boolean,
  workspace: string,
  fileIds: string[],
  supportsSelfTest: boolean,
  supportsReset: boolean,
  id: string,
  location: AssetLocationModel,
  calibrationStatus: 'OK' | 'APPROACHING_RECOMMENDED_DUE_DATE' | 'PAST_RECOMMENDED_DUE_DATE' | string,
  isSystemController: boolean,
  externalCalibration?: ExternalCalibrationModel,
  discoveryType: 'MANUAL' | 'AUTOMATIC' | string,
  properties: Record<string, string>,
  keywords: string[],
  lastUpdatedTimestamp: string,
}

export interface AssetPresenceWithSystemConnectionModel {
  assetPresence: "INITIALIZING" | "UNKNOWN" | "NOT_PRESENT" | "PRESENT" | string,
  // to be compatible with both SLS and SLE
  systemConnection?: "APPROVED" | "DISCONNECTED" | "CONNECTED_UPDATE_PENDING" | "CONNECTED_UPDATE_SUCCESSFUL" | "CONNECTED_UPDATE_FAILED" | "UNSUPPORTED" | "ACTIVATED" | "CONNECTED" | string
}

export interface AssetLocationModel {
  minionId: string,
  parent: string,
  resourceUri: string,
  slotNumber: number,
  state: AssetPresenceWithSystemConnectionModel,
  physicalLocation?: string
}

export interface ExternalCalibrationModel {
  temperatureSensors: any[],
  isLimited?: boolean,
  date: string,
  recommendedInterval: number,
  nextRecommendedDate: string,
  nextCustomDueDate?: string,
  comments: string,
  entryType: "AUTOMATIC" | "MANUAL" | string,
  operator: ExternalCalibrationOperatorModel
}

export interface ExternalCalibrationOperatorModel {
  displayName: string;
  userId: string
}

export enum AssetFilterProperties {
  AssetIdentifier = 'AssetIdentifier',
  SerialNumber = 'SerialNumber',
  ModelName = 'ModelName',
  VendorName = 'VendorName',
  VendorNumber = 'VendorNumber',
  AssetName = 'AssetName',
  FirmwareVersion = 'FirmwareVersion',
  HardwareVersion = 'HardwareVersion',
  BusType = 'BusType',
  IsNIAsset = 'IsNIAsset',
  Keywords = 'Keywords',
  Properties = 'Properties',
  LocationMinionId = 'Location.MinionId',
  LocationSlotNumber = 'Location.SlotNumber',
  LocationAssetStateSystemConnection = 'Location.AssetState.SystemConnection',
  LocationAssetStateAssetPresence = 'Location.AssetState.AssetPresence',
  SupportsSelfCalibration = 'SupportsSelfCalibration',
  SelfCalibrationCalibrationDate = 'SelfCalibration.CalibrationDate',
  SupportsExternalCalibration = 'SupportsExternalCalibration',
  CustomCalibrationInterval = 'CustomCalibrationInterval',
  CalibrationStatus = 'CalibrationStatus',
  ExternalCalibrationCalibrationDate = 'ExternalCalibration.CalibrationDate',
  ExternalCalibrationNextRecommendedDate = 'ExternalCalibration.NextRecommendedDate',
  ExternalCalibrationRecommendedInterval = 'ExternalCalibration.RecommendedInterval',
  ExternalCalibrationComments = 'ExternalCalibration.Comments',
  ExternalCalibrationIsLimited = 'ExternalCalibration.IsLimited',
  ExternalCalibrationOperatorDisplayName = 'ExternalCalibration.Operator.DisplayName',
  IsSystemController = 'IsSystemController'
}

export interface CalibrationForecastResponse {
  calibrationForecast: CalibrationForecastModel
}

export interface CalibrationForecastModel {
  columns: FieldDTO[],
}
