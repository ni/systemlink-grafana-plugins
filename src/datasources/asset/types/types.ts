import { DataSourceJsonData } from "@grafana/data";
import { DataQuery } from "@grafana/schema";


export enum AssetQueryType {
  None = "",
  ListAssets = "List Assets",
  CalibrationForecast = "Calibration Forecast",
  AssetSummary = "Asset Summary"
}

export interface AssetQuery extends DataQuery {
  type: AssetQueryType
}

export interface AssetFeatureToggles {
  calibrationForecast: boolean;
  assetList: boolean;
  assetSummary: boolean;
}

export interface AssetDataSourceOptions extends DataSourceJsonData {
  featureToggles: AssetFeatureToggles;
}

export const AssetFeatureTogglesDefaults: AssetFeatureToggles = {
  assetList: true,
  calibrationForecast: true,
  assetSummary: true
}

export enum AssetQueryReturnType {
  AssetIdentification = 'Asset Identification',
  AssetId = 'Asset Id'
}

export enum BusType {
  BUILT_IN_SYSTEM = 'BUILT_IN_SYSTEM',
  PCI_PXI = 'PCI_PXI',
  USB = 'USB',
  GPIB = 'GPIB',
  VXI = 'VXI',
  SERIAL = 'SERIAL',
  TCP_IP = 'TCP_IP',
  CRIO = 'CRIO',
  SCXI = 'SCXI',
  CDAQ = 'CDAQ',
  SWITCH_BLOCK = 'SWITCH_BLOCK',
  SCC = 'SCC',
  FIRE_WIRE = 'FIRE_WIRE',
  ACCESSORY = 'ACCESSORY',
  CAN = 'CAN',
  SWITCH_BLOCK_DEVICE = 'SWITCH_BLOCK_DEVICE',
  SLSC = 'SLSC'
}

export const BusTypeOptions = [
  { label: 'Built-in-system', value: BusType.BUILT_IN_SYSTEM },
  { label: 'PCI/PXI', value: BusType.PCI_PXI },
  { label: 'USB', value: BusType.USB },
  { label: 'GPIB', value: BusType.GPIB },
  { label: 'VXI', value: BusType.VXI },
  { label: 'Serial', value: BusType.SERIAL },
  { label: 'TCP/IP', value: BusType.TCP_IP },
  { label: 'CompactRIO', value: BusType.CRIO },
  { label: 'SCXI', value: BusType.SCXI },
  { label: 'cDAQ', value: BusType.CDAQ },
  { label: 'SwitchBlock', value: BusType.SWITCH_BLOCK },
  { label: 'SCC', value: BusType.SCC },
  { label: 'FireWire', value: BusType.FIRE_WIRE },
  { label: 'ACCESSORY', value: BusType.ACCESSORY },
  { label: 'CAN', value: BusType.CAN },
  { label: 'SwitchBlock device', value: BusType.SWITCH_BLOCK_DEVICE },
  { label: 'SLSC', value: BusType.SLSC },
];

export enum AssetType {
  GENERIC = 'GENERIC',
  DEVICE_UNDER_TEST = 'DEVICE_UNDER_TEST',
  FIXTURE = 'FIXTURE',
  SYSTEM = 'SYSTEM'
};

export const AssetTypeOptions = [
  { label: 'Generic', value: AssetType.GENERIC },
  { label: 'Device under test', value: AssetType.DEVICE_UNDER_TEST },
  { label: 'Fixture', value: AssetType.FIXTURE },
  { label: 'System', value: AssetType.SYSTEM },
];
