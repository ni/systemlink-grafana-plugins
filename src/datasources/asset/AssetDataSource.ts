import {
  DataFrameDTO,
  DataQueryRequest,
  DataSourceInstanceSettings,
  TestDataSourceResponse,
} from '@grafana/data';
import { BackendSrv, getBackendSrv, getTemplateSrv, TemplateSrv } from '@grafana/runtime';
import { DataSourceBase } from 'core/DataSourceBase';
import {
  AssetFilterProperties,
  AssetMetadataQuery,
  AssetModel,
  AssetsResponse,
} from './types';
import { getWorkspaceName, replaceVariables } from "../../core/utils";
import { SystemMetadata } from "../system/types";
import { defaultOrderBy, defaultProjection } from "../system/constants";

export class AssetDataSource extends DataSourceBase<AssetMetadataQuery> {
  constructor(
    readonly instanceSettings: DataSourceInstanceSettings,
    readonly backendSrv: BackendSrv = getBackendSrv(),
    readonly templateSrv: TemplateSrv = getTemplateSrv()
  ) {
    super(instanceSettings, backendSrv, templateSrv);
  }

  baseUrl = this.instanceSettings.url + '/niapm/v1';

  defaultQuery = {
    workspace: '',
    minionIds: [],
    groupBy: []
  };

  async runQuery(query: AssetMetadataQuery, options: DataQueryRequest): Promise<DataFrameDTO> {
    return this.processMetadataQuery(query as AssetMetadataQuery);
  }

  async processMetadataQuery(query: AssetMetadataQuery) {
    const result: DataFrameDTO = { refId: query.refId, fields: [] };
    const minionIds = replaceVariables(query.minionIds, this.templateSrv);
    let workspaceId = this.templateSrv.replace(query.workspace);
    const conditions = [];
    if (minionIds.length) {
      const systemsCondition = minionIds.map(id => `${AssetFilterProperties.LocationMinionId} = "${id}"`)
      conditions.push(`(${systemsCondition.join(' or ')})`);
    }
    if (workspaceId) {
      conditions.push(`workspace = "${workspaceId}"`);
    }
    const assetFilter = conditions.join(' and ');
    const assets: AssetModel[] = await this.queryAssets(assetFilter, 1000);
    const workspaces = await this.getWorkspaces();
    result.fields = [
      { name: 'id', values: assets.map(a => a.id) },
      { name: 'name', values: assets.map(a => a.name) },
      { name: 'model name', values: assets.map(a => a.modelName) },
      { name: 'serial number', values: assets.map(a => a.serialNumber) },
      { name: 'bus type', values: assets.map(a => a.busType) },
      { name: 'asset type', values: assets.map(a => a.assetType) },
      { name: 'is NI asset', values: assets.map(a => a.isNIAsset) },
      { name: 'calibration status', values: assets.map(a => a.calibrationStatus) },
      { name: 'is system controller', values: assets.map(a => a.isSystemController) },
      { name: 'last updated timestamp', values: assets.map(a => a.lastUpdatedTimestamp) },
      { name: 'minionId', values: assets.map(a => a.location.minionId) },
      { name: 'parent name', values: assets.map(a => a.location.parent) },
      { name: 'workspace', values: assets.map(a => getWorkspaceName(workspaces, a.workspace)) },
    ];
    return result;
  }

  shouldRunQuery(_: AssetMetadataQuery): boolean {
    return true;
  }

  async queryAssets(filter = '', take = -1): Promise<AssetModel[]> {
    let data = { filter, take };
    try {
      let response = await this.post<AssetsResponse>(this.baseUrl + '/query-assets', data);
      return response.assets;
    } catch (error) {
      throw new Error(`An error occurred while querying assets: ${error}`);
    }
  }

  async querySystems(filter = '', projection = defaultProjection): Promise<SystemMetadata[]> {
    try {
      let response = await this.getSystems({
        filter: filter,
        projection: `new(${projection.join()})`,
        orderBy: defaultOrderBy,
      })

      return response.data;
    } catch (error) {
      throw new Error(`An error occurred while querying systems: ${error}`);
    }
  }

  async testDatasource(): Promise<TestDataSourceResponse> {
    await this.get(this.baseUrl + '/assets?take=1');
    return { status: 'success', message: 'Data source connected and authentication successful!' };
  }
}
