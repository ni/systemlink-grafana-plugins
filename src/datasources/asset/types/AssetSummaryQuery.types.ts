import { DataQuery } from '@grafana/schema'
import { AssetQueryType } from './types';

export interface AssetSummaryQuery extends DataQuery {
  queryType: AssetQueryType;
}

export interface AssetSummaryResponse extends DataQuery {
  total: number;
  active: number;
  notActive: number;
  approachingRecommendedDueDate: number;
  pastRecommendedDueDate: number;
}

