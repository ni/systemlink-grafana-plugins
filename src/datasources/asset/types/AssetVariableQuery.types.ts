import { AssetQuery } from './types';

export interface AssetVariableQuery extends AssetQuery {
  filter: string;
  take?: number;
}
