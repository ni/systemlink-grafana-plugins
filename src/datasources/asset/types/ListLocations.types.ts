export interface ListLocationsResponse {
    locations: LocationModel[];
}

export interface LocationModel {
    id: string;
    name: string;
}
