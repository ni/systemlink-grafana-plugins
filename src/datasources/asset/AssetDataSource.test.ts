import { BackendSrv } from "@grafana/runtime";
import { MockProxy } from "jest-mock-extended";
import {
  createFetchError,
  createFetchResponse,
  getQueryBuilder,
  requestMatching,
  setupDataSource,
} from "test/fixtures";
import { AssetDataSource } from "./AssetDataSource";
import { AssetQueryType } from "./types/types";
import { AssetPresenceWithSystemConnectionModel, AssetsResponse } from "datasources/asset-common/types";
import { ListAssetsQuery } from "./types/ListAssets.types";
import { AssetVariableQuery } from "./types/AssetVariableQuery.types";

let ds: AssetDataSource, backendSrv: MockProxy<BackendSrv>
let assetOptions = {
  featureToggles: {
    assetListEnabled: true,
    calibrationForecastEnabled: true,
    assetSummaryEnabled: true,
  }
}

beforeEach(() => {
  [ds, backendSrv] = setupDataSource(AssetDataSource, () => assetOptions);
});

const assetsResponseMock: AssetsResponse =
{
  "assets": [
    {
      "modelName": "sbRIO-9629",
      "modelNumber": 31299,
      "serialNumber": "01FE20D1",
      "vendorName": "National Instruments",
      "vendorNumber": 0,
      "busType": "BUILT_IN_SYSTEM",
      "name": "NI-sbRIO-9629-01FE20D1",
      "assetType": "SYSTEM",
      "discoveryType": "AUTOMATIC",
      "firmwareVersion": "8.8.0f0",
      "hardwareVersion": "",
      "partNumber": "1",
      "visaResourceName": "",
      "temperatureSensors": [
        {
          "name": "FPGA",
          "reading": 38.5625
        },
        {
          "name": "Primary",
          "reading": 36.75
        },
        {
          "name": "Secondary",
          "reading": 34.8125
        },
        {
          "name": "CPU Core 2",
          "reading": 43
        },
        {
          "name": "CPU Core 3",
          "reading": 43
        },
        {
          "name": "CPU Core 0",
          "reading": 43
        },
        {
          "name": "CPU Core 1",
          "reading": 43
        }
      ],
      "supportsSelfCalibration": true,
      "selfCalibration": {
        "date": "2024-02-21T12:54:20.069Z",
      },
      "supportsExternalCalibration": false,
      "isNIAsset": true,
      "id": "7f6d0d74-bc75-4d78-9edd-00c253b3a0de",
      "location": {
        "minionId": "NI_sbRIO-9629--SN-01FE20D1--MAC-00-80-2F-33-30-18",
        "parent": "",
        "resourceUri": "system",
        "slotNumber": -1,
        "state": {
          "assetPresence": "PRESENT"
        }
      },
      "calibrationStatus": "OK",
      "isSystemController": true,
      "workspace": "e73fcd94-649b-4d0a-b164-bf647a5d0946",
      "properties": { "Property1": "Test1", "Property2": "Test2" },
      "keywords": ["TestKeyword1", "TestKeyword2"],
      "lastUpdatedTimestamp": "2024-02-21T12:54:20.069Z",
      "fileIds": [],
      "supportsSelfTest": false,
      "supportsReset": false
    },
    {
      "modelName": "AI 0-15",
      "modelNumber": 31303,
      "serialNumber": "01FE20D1",
      "vendorName": "National Instruments",
      "vendorNumber": 4243,
      "busType": "CRIO",
      "partNumber": "2",
      "name": "Conn0_AI",
      "assetType": "GENERIC",
      "discoveryType": "AUTOMATIC",
      "firmwareVersion": "",
      "hardwareVersion": "B",
      "visaResourceName": "PXI::203.0::INSTR",
      "temperatureSensors": [],
      "supportsSelfCalibration": false,
      "supportsExternalCalibration": true,
      "isNIAsset": true,
      "id": "71fbef61-fae6-4364-82d4-29feb79c7146",
      "location": {
        "minionId": "NI_sbRIO-9629--SN-01FE20D1--MAC-00-80-2F-33-30-18",
        "parent": "sbRIO1",
        "resourceUri": "7/4243/31303/01FE20D1/3",
        "slotNumber": 3,
        "state": {
          "assetPresence": "PRESENT",
          "systemConnection": "DISCONNECTED"
        } as AssetPresenceWithSystemConnectionModel
      },
      "calibrationStatus": "OK",
      "isSystemController": false,
      "workspace": "e73fcd94-649b-4d0a-b164-bf647a5d0946",
      "properties": {},
      "keywords": [ "Kw1", "Kw2" ],
      "lastUpdatedTimestamp": "2024-02-21T12:54:20.072Z",
      "fileIds": [],
      "supportsSelfTest": false,
      "supportsReset": false
    },
    {
      "modelName": "AO 0-3",
      "modelNumber": 31304,
      "serialNumber": "01FE20D1",
      "vendorName": "National Instruments",
      "vendorNumber": 4243,
      "busType": "CRIO",
      "partNumber": "3",
      "name": "Conn0_AO",
      "assetType": "GENERIC",
      "discoveryType": "AUTOMATIC",
      "firmwareVersion": "",
      "hardwareVersion": "B",
      "visaResourceName": "",
      "temperatureSensors": [],
      "supportsSelfCalibration": false,
      "supportsExternalCalibration": true,
      "isNIAsset": true,
      "id": "cf1ac843-06a2-4713-ab43-f9d1d8dfdc32",
      "location": {
        "minionId": "NI_sbRIO-9629--SN-01FE20D1--MAC-00-80-2F-33-30-18",
        "parent": "sbRIO1",
        "resourceUri": "7/4243/31304/01FE20D1/4",
        "slotNumber": 4,
        "state": {
          "assetPresence": "PRESENT",
          "systemConnection": "DISCONNECTED"
        } as AssetPresenceWithSystemConnectionModel
      },
      "calibrationStatus": "OK",
      "isSystemController": false,
      "workspace": "e73fcd94-649b-4d0a-b164-bf647a5d0946",
      "properties": {},
      "keywords": [],
      "lastUpdatedTimestamp": "2024-02-21T12:54:20.076Z",
      "fileIds": [],
      "supportsSelfTest": false,
      "supportsReset": false
    },
    {
      "modelName": "DIO 0-3",
      "modelNumber": 31305,
      "serialNumber": "01FE20D1",
      "vendorName": "National Instruments",
      "vendorNumber": 4243,
      "busType": "CRIO",
      "name": "Conn0_DIO0-3",
      "assetType": "GENERIC",
      "discoveryType": "AUTOMATIC",
      "firmwareVersion": "",
      "hardwareVersion": "",
      "partNumber": "",
      "visaResourceName": "",
      "temperatureSensors": [],
      "supportsSelfCalibration": false,
      "supportsExternalCalibration": false,
      "isNIAsset": true,
      "id": "456e8812-1da4-4818-afab-f0cd34f74567",
      "location": {
        "minionId": "NI_sbRIO-9629--SN-01FE20D1--MAC-00-80-2F-33-30-18",
        "physicalLocation": "Cabinet 1",
        "parent": "sbRIO1",
        "resourceUri": "7/4243/31305/01FE20D1/5",
        "slotNumber": 5,
        "state": {
          "assetPresence": "PRESENT",
          "systemConnection": "DISCONNECTED"
        } as AssetPresenceWithSystemConnectionModel
      },
      "calibrationStatus": "OK",
      "isSystemController": false,
      "workspace": "e73fcd94-649b-4d0a-b164-bf647a5d0946",
      "properties": {},
      "keywords": [],
      "lastUpdatedTimestamp": "2024-02-21T12:54:20.078Z",
      "fileIds": [],
      "supportsSelfTest": true,
      "supportsReset": true,
      "externalCalibration": {
        "temperatureSensors": [],
        "isLimited": false,
        "date": "2024-02-21T12:54:20.078Z",
        "recommendedInterval": 12,
        "nextRecommendedDate": "2025-02-21T12:54:20.078Z",
        "nextCustomDueDate": "2025-02-21T12:54:20.078Z",
        "resolvedDueDate": "2025-02-21T12:54:20.078Z",
        "entryType": "AUTOMATIC",
        "updatedBy": "user-id"
      }
    }
  ],
  "totalCount": 4
}

const assetsWithoutNameResponseMock: AssetsResponse =
{
  "assets": [
    {
      "modelName": "sbRIO-9629",
      "modelNumber": 0,
      "serialNumber": "01FE20D1",
      "vendorName": "National Instruments",
      "vendorNumber": 0,
      "busType": "BUILT_IN_SYSTEM",
      "name": "",
      "assetType": "SYSTEM",
      "discoveryType": "AUTOMATIC",
      "firmwareVersion": "8.8.0f0",
      "hardwareVersion": "",
      "partNumber": "4",
      "visaResourceName": "",
      "temperatureSensors": [],
      "supportsSelfCalibration": false,
      "supportsExternalCalibration": false,
      "isNIAsset": true,
      "id": "7f6d0d74-bc75-4d78-9edd-00c253b3a0de",
      "location": {
        "minionId": "NI_sbRIO-9629--SN-01FE20D1--MAC-00-80-2F-33-30-18",
        "parent": "",
        "resourceUri": "system",
        "slotNumber": -1,
        "state": {
          "assetPresence": "PRESENT"
        }
      },
      "calibrationStatus": "OK",
      "isSystemController": true,
      "workspace": "e73fcd94-649b-4d0a-b164-bf647a5d0946",
      "properties": {},
      "keywords": [],
      "lastUpdatedTimestamp": "2024-02-21T12:54:20.069Z",
      "fileIds": [],
      "supportsSelfTest": false,
      "supportsReset": false
    },
    {
      "modelName": "",
      "modelNumber": 1111,
      "serialNumber": "2222",
      "vendorName": "",
      "vendorNumber": 3333,
      "busType": "BUILT_IN_SYSTEM",
      "name": "",
      "assetType": "SYSTEM",
      "discoveryType": "AUTOMATIC",
      "firmwareVersion": "8.8.0f0",
      "hardwareVersion": "",
      "partNumber": "5",
      "visaResourceName": "",
      "temperatureSensors": [],
      "supportsSelfCalibration": false,
      "supportsExternalCalibration": false,
      "isNIAsset": true,
      "id": "7f6d0d74-bc75-4d78-9edd-00c253b3a0de",
      "location": {
        "minionId": "NI_sbRIO-9629--SN-01FE20D1--MAC-00-80-2F-33-30-18",
        "parent": "",
        "resourceUri": "system",
        "slotNumber": -1,
        "state": {
          "assetPresence": "PRESENT"
        }
      },
      "calibrationStatus": "OK",
      "isSystemController": true,
      "workspace": "e73fcd94-649b-4d0a-b164-bf647a5d0946",
      "properties": {},
      "keywords": [],
      "lastUpdatedTimestamp": "2024-02-21T12:54:20.069Z",
      "fileIds": [],
      "supportsSelfTest": false,
      "supportsReset": false
    }
  ],
  "totalCount": 2
}


const assetMetadataQueryMock: ListAssetsQuery = {
  type: AssetQueryType.ListAssets,
  filter: 'Location.MinionId == "123"',
  refId: ''
}

const buildMetadataQuery = getQueryBuilder<ListAssetsQuery>()({
  filter: ''
});

describe('testDatasource', () => {
  test('returns success', async () => {

    backendSrv.fetch
      .calledWith(requestMatching({ url: '/niapm/v1/assets?take=1' }))
      .mockReturnValue(createFetchResponse(25));

    const result = await ds.testDatasource();

    expect(result.status).toEqual('success');
  });

  test('bubbles up exception', async () => {
    backendSrv.fetch
      .calledWith(requestMatching({ url: '/niapm/v1/assets?take=1' }))
      .mockReturnValue(createFetchError(400));

    await expect(ds.testDatasource()).rejects.toThrow('Request to url "/niapm/v1/assets?take=1" failed with status code: 400. Error message: "Error"');
  });
})

describe('queries', () => {
  test('run metadata query', async () => {
    backendSrv.fetch
      .calledWith(requestMatching({ url: '/niapm/v1/query-assets' }))
      .mockReturnValue(createFetchResponse(assetsResponseMock as AssetsResponse))

    const result = await ds.query(buildMetadataQuery(assetMetadataQueryMock))

    expect(result.data).toMatchSnapshot()
  })

  test('handles metadata query error', async () => {
    backendSrv.fetch
      .calledWith(requestMatching({ url: '/niapm/v1/query-assets' }))
      .mockReturnValue(createFetchError(418))

    await expect(ds.query(buildMetadataQuery(assetMetadataQueryMock))).rejects.toThrow()
  })

  describe('metricFindQuery', () => {
    it('returns name/alias when asset name field is present', async () => {
      const query: AssetVariableQuery = {
        filter: '',
        type: AssetQueryType.None,
        refId: ""
      }

      backendSrv.fetch
        .calledWith(requestMatching({ url: '/niapm/v1/query-assets' }))
        .mockReturnValue(createFetchResponse(assetsResponseMock as AssetsResponse))

      const options = {
        scopedVars: {}
      }

      const result = await ds.metricFindQuery(query, options)

      expect(result).toMatchSnapshot();
    })

    it('returns default identifiers when asset name is not present', async () => {
      const query: AssetVariableQuery = {
        filter: '',
        type: AssetQueryType.None,
        refId: ""
      }

      backendSrv.fetch
        .calledWith(requestMatching({ url: '/niapm/v1/query-assets' }))
        .mockReturnValue(createFetchResponse(assetsWithoutNameResponseMock as AssetsResponse))

      const options = {
        scopedVars: {}
      }

      const result = await ds.metricFindQuery(query, options)

      expect(result).toMatchSnapshot();
    })
  })
})
