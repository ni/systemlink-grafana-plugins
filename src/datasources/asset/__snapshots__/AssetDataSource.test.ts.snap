// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries metricFindQuery returns default identifiers when asset name is not present 1`] = `
[
  {
    "text": "Vendor: National Instruments - Model: sbRIO-9629 - Serial: 01FE20D1",
    "value": "Assets.National Instruments.sbRIO-9629.01FE20D1",
  },
  {
    "text": "Vendor: 3333 - Model: 1111 - Serial: 2222",
    "value": "Assets.3333.1111.2222",
  },
]
`;

exports[`queries metricFindQuery returns name/alias when asset name field is present 1`] = `
[
  {
    "text": "NI-sbRIO-9629-01FE20D1 (Vendor: National Instruments - Model: sbRIO-9629 - Serial: 01FE20D1)",
    "value": "Assets.National Instruments.sbRIO-9629.01FE20D1",
  },
  {
    "text": "Conn0_AI (Vendor: National Instruments - Model: AI 0-15 - Serial: 01FE20D1)",
    "value": "Assets.National Instruments.AI 0-15.01FE20D1",
  },
  {
    "text": "Conn0_AO (Vendor: National Instruments - Model: AO 0-3 - Serial: 01FE20D1)",
    "value": "Assets.National Instruments.AO 0-3.01FE20D1",
  },
  {
    "text": "Conn0_DIO0-3 (Vendor: National Instruments - Model: DIO 0-3 - Serial: 01FE20D1)",
    "value": "Assets.National Instruments.DIO 0-3.01FE20D1",
  },
]
`;

exports[`queries run metadata query 1`] = `
[
  {
    "fields": [
      {
        "name": "id",
        "values": [
          "7f6d0d74-bc75-4d78-9edd-00c253b3a0de",
          "71fbef61-fae6-4364-82d4-29feb79c7146",
          "cf1ac843-06a2-4713-ab43-f9d1d8dfdc32",
          "456e8812-1da4-4818-afab-f0cd34f74567",
        ],
      },
      {
        "name": "name",
        "values": [
          "NI-sbRIO-9629-01FE20D1",
          "Conn0_AI",
          "Conn0_AO",
          "Conn0_DIO0-3",
        ],
      },
      {
        "name": "model name",
        "values": [
          "sbRIO-9629",
          "AI 0-15",
          "AO 0-3",
          "DIO 0-3",
        ],
      },
      {
        "name": "serial number",
        "values": [
          "01FE20D1",
          "01FE20D1",
          "01FE20D1",
          "01FE20D1",
        ],
      },
      {
        "name": "bus type",
        "values": [
          "BUILT_IN_SYSTEM",
          "CRIO",
          "CRIO",
          "CRIO",
        ],
      },
      {
        "name": "asset type",
        "values": [
          "SYSTEM",
          "GENERIC",
          "GENERIC",
          "GENERIC",
        ],
      },
      {
        "name": "is NI asset",
        "values": [
          true,
          true,
          true,
          true,
        ],
      },
      {
        "name": "calibration status",
        "values": [
          "OK",
          "OK",
          "OK",
          "OK",
        ],
      },
      {
        "name": "is system controller",
        "values": [
          true,
          false,
          false,
          false,
        ],
      },
      {
        "name": "last updated timestamp",
        "values": [
          "2024-02-21T12:54:20.069Z",
          "2024-02-21T12:54:20.072Z",
          "2024-02-21T12:54:20.076Z",
          "2024-02-21T12:54:20.078Z",
        ],
      },
      {
        "name": "location",
        "values": [
          "",
          "",
          "",
          "Cabinet 1",
        ],
      },
      {
        "name": "minionId",
        "values": [
          "NI_sbRIO-9629--SN-01FE20D1--MAC-00-80-2F-33-30-18",
          "NI_sbRIO-9629--SN-01FE20D1--MAC-00-80-2F-33-30-18",
          "NI_sbRIO-9629--SN-01FE20D1--MAC-00-80-2F-33-30-18",
          "NI_sbRIO-9629--SN-01FE20D1--MAC-00-80-2F-33-30-18",
        ],
      },
      {
        "name": "parent name",
        "values": [
          "",
          "sbRIO1",
          "sbRIO1",
          "sbRIO1",
        ],
      },
      {
        "name": "workspace",
        "values": [
          "e73fcd94-649b-4d0a-b164-bf647a5d0946",
          "e73fcd94-649b-4d0a-b164-bf647a5d0946",
          "e73fcd94-649b-4d0a-b164-bf647a5d0946",
          "e73fcd94-649b-4d0a-b164-bf647a5d0946",
        ],
      },
      {
        "name": "calibration due date",
        "values": [
          undefined,
          undefined,
          undefined,
          "2025-02-21T12:54:20.078Z",
        ],
      },
    ],
    "refId": "A",
  },
]
`;
